Spark: Compact Medium Size Files (#460)

RussellSpitzer opened the pull request:
Previously the rewrite action would ignore files which were below our
target threshold if we could not combine them with a smaller file or piece
of a larger file. This would preclude ever combining the piece of of medium
sized file with several other medium sized files resulting in files of the
correct size. 

Example: Say our target is 100MB, and our files are all 75MB with 25MB Splits
Previously we would consider all 75MB chunks of data as un-splitable. This means we could never optimize the files
together. So if for example we had 4 such files
`(75, 75, 75, 75)` we could never make these four files into 3 `(100, 100, 100)`
Since there are no two whole files that can be combined to make a target size file `(75 + 75) = 150`
even though we obviously could by dividing one of the 75 MB files up into 25 MB pieces and spreading it's data over the other files.

To fix this we attempt to split up all files we are compacting at the planning phase
to give a chance for pieces of medium sized files to be incorporated with each
other (or with small files).

RussellSpitzer commented:
@jackye1995 + @rdblue - If you would like to take a look

RussellSpitzer commented:
@rdblue fixed up and rebased

RussellSpitzer commented:
@rdblue if this is good to go I can make the identical changes to the other Spark modules and merge

+ @zhangjun0x01 You may want to take a look at this as well

aokolnychyi commented on spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkFilesScan.java:
```diff
@@ -37,7 +37,8 @@
 class SparkFilesScan extends SparkBatchScan {
   private final String taskSetID;
-  private final long splitSize;
+  private final Long readSplitSize;
```
nit: why use boxed values if methods actually return primitives?

aokolnychyi commented on spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:
```diff
@@ -57,6 +57,9 @@ private SparkReadOptions() {
   public static final String FILE_SCAN_TASK_SET_ID = "file-scan-task-set-id";
 
+  // Set the target task size of a file scan combined tasks
+  public static final String FILE_SCAN_TARGET_SIZE = "file-scan-target-size";
```
The name is not very descriptive, to be honest. It starts with `file-scan`, which makes it specific to `SparkFilesScan`, like it should but it is not clear what `target-size` means. I think it defines the target task size whenever reading file scan task sets.

I don't really have a good alternative to suggest and it is an internal option, so I probably wouldn't worry.

RussellSpitzer replied:
It is internal so I don't really want to worry about it too much

aokolnychyi commented on spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3BinPackStrategy.java:
```diff
@@ -61,9 +62,17 @@ public Table table() {
+      long targetReadSize = splitSize(inputFileSize(filesToRewrite));
+      // Ideally this would be the row-group size but the row group size is not guaranteed to be consistent
+      long fileSplitSize = Long.valueOf(table.properties().getOrDefault(
+          TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES,
```
It looks a bit odd to simply check the row group config as the table may have both Avro and Parquet and we may be compacting Avro, for example. I know it is something @rdblue has suggested so I'll leave it to you folks.

aokolnychyi reviewed (approved):
Looks good to me.

aokolnychyi commented:
@rdblue, could you take a look? You left comments and I am not sure they are resolved.

jackye1995 reviewed (commented):
Overall I think this makes sense to me to separate these 2 concepts. For determining the `fileSplitSize`, I think we should only use `PARQUET_ROW_GROUP_SIZE_BYTES` when the table's file format is Parquet (although in the future people might have mixture, but this is our best effort).

rdblue commented on spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java:
```diff
@@ -155,6 +155,18 @@ public long splitOpenFileCost() {
+   * for a different configuration than split size. Having this larger than split size allows files
+   * to be cut into parts and then recombined in an attempt to get closer to the target size.
+   */
+  public long planTargetSize() {
```
Is it possible to use `targetSize` from the scan like we just updated #3400 to do?

rdblue replied:
Nevermind, this looks different. Should this use the table property `write.target-file-size-bytes`?

rdblue commented:
@RussellSpitzer, if I understand what this is doing correctly, it looks like this is fixing two problems. First, Iceberg can probably pack tasks more densely at read time. The 75/75/75/75 situation you describe also applies to regular task planning and that's why you're changing the config there. Second, the bin packing strategy needs to configure the scan it creates with the new config you introduced. But, this has to guess at how it can split files.

I think that we may be able to do this in a more straightforward way. Instead of using two configs, a size to split tasks down to and a size to combine up to, I think we can update just the split code to be a little smarter.

We already have a `split` implementation of `FileScanTask` that uses row group offsets. That's the smallest chunk of a Parquet or ORC file that we can split down to. What if we produce one task per row group rather than trying to guess how far down to split? Then we could combine just like normal. We could also avoid producing a ton of tiny splits by basing this on a heuristic and the original target split size. In `FileScanTask.split`, actually split down to `targetSize / 4`.

What do you think? That would simplify configuration and make it possible for us to more densely pack splits for normal scans as well as those for compaction.

If you like it, I think we would also want to add a way to combine tasks for the same file back into larger splits. For example, `FileScanTask(a.parquet, 0, 50)` and `FileScanTask(a.parquet, 50, 100)` in the same combined task would be rewritten as `FileScanTask(a.parquet, 0, 100)`.

RussellSpitzer commented:
@rdblue I think that's a good idea but it may take a bit of work. Currently you always have to pass through "something" to splitSize although I think passing through 1 as split size would effectively only return one split per row group. The issue is still with Avro where we have no idea and the 1 ends up being a bad idea.

Let me think about this ...

RussellSpitzer commented:
I'm going to try out something like

```java

  public static CloseableIterable<FileScanTask> splitOnRowGroups(CloseableIterable<FileScanTask> tasks, long fallbackSplitSize) {
    Iterable<FileScanTask> splitTasks = FluentIterable
        .from(tasks)
        .transformAndConcat(input -> {
          if (input.file().splitOffsets() != null) {
            return input.split(0); // Split on offsets, 1 Offset per split
          } else {
            return input.split(fallbackSplitSize);
          }
        });
    // Capture manifests which can be closed after scan planning
    return CloseableIterable.combine(splitTasks, tasks);
  }
  ```

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:
```diff
@@ -51,6 +51,31 @@ public static boolean hasDeletes(FileScanTask task) {
+   * @param fallbackSplitSize the splitSize to use when the file does not contain explicit offsets to use
+   * @return Scan tasks, one per offset
+   */
+  public static CloseableIterable<FileScanTask> splitOnOffsets(CloseableIterable<FileScanTask> tasks,
```
@rdblue Changed implementation here, didn't recombine splits for the same file yet, but wanted to make sure you are ok with this. We could theoretically change this for all splitFiles but I didn't want to change that api yet ...

rdblue commented on api/src/main/java/org/apache/iceberg/FileFormat.java:
```diff
@@ -25,23 +25,29 @@
   }
 
+  public boolean hasOffsets() {
+    return offsets;
```
Don't we want to treat every file the same? If it has split offsets, we use them and if not we try to split at the target size? That is dependent on the file metadata and not the format.

RussellSpitzer replied:
This seemed much more confusing to me because of the split behavior on files without split offset information. I'd rather it's just explicit, we have files we know how to split but can only be split on offsets, and we have files that we know how to split can be split arbitrarily. I was trying to figure out a another term for this, basically I want to differentiate between formats like ORC , Parquet and Avro

RussellSpitzer replied:
To be clear, this is not treating every file the same but saying that this file has the potential of having offsets as opposed to being arbitrarily splittable.

rdblue commented on core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:
```diff
@@ -61,6 +62,39 @@ public static boolean hasDeletes(FileScanTask task) {
+              // Split on offsets, size 0 means 1 task per offset
+              return input.split(0);
+            } else {
+              // File too small to have offsets, take the entire file as a task
```
Seems like we could just use the original split method here, just in case there are no split locations (which are optional) but the file isn't small. I don't think we need to make this assumption.

RussellSpitzer replied:
This is again where I wanted to differentiate between globally splittable and only splittable on offsets. Originally I just used the base method but then I have to pass through the file size to get a single split which seemed odd. We have to pass through the size because we go through the non-offset split path which can possibly give us multiple splits (some of which would be empty.)

jackye1995 replied:
Just want to combine the original split code here to have a general view, basically we get:

```java
if (file.format().hasOffsets()) {
    if (file.splitOffsets() != null) {
        // split to 1 task per offset
    } else {
       // file too small, use 1 task for entire file
    }
} else {
    if (file.format().isSplittable()) {
        if (file.splitOffsets() != null) {
            // offset aware split
        } else {
            // fixed size split
        }
    } else {
        // use 1 task for entire file
    }
}
```

So the case of `hasOffset && splitOffsets == null` case is basically a short-circuit even if the file is splittable, there is still some value for that.

The question I have is that would it be worth directly adding the logic to the original splitFiles, instead of having 2 different methods? I imagine this would also help reducing the number of file scan tasks generated during scan planning.

RussellSpitzer replied:
I think it can be a little simpler
```java
if (file.format().hasOffsets()) {
    if (file.splitOffsets() != null) {
        // split to 1 task per offset
    } else {
       // file too small, use 1 task for entire file
    }
} else {
    if (file.format().isSplittable()) {
       // fixed size split (this cannot be an offset split because that would have been in the first branch)
      }
    } else {
        // use 1 task for entire file
    }
}
```

But yeah I think we can have this in the original logic, I was just afraid of changing an existing behavior mostly.

rdblue replied:
I think that logic is incorrect. If a file doesn't have offsets, then we can't assume that it is small. It could just be that a writer didn't produce the offsets. The Parquet import code may or may not produce offsets, for example. We can't assume something about the file just because metadata is missing.

RussellSpitzer replied:
I decided to forgo this route and try something we were discussing a bit earlier. Instead of trying to split offset files exactly on split-size boundaries, I'm going to split all files with offsets on their offsets and then recombine them when they are put in a CombinedScanTask.

jackye1995 commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3BinPackStrategy.java:
```diff
@@ -64,6 +64,7 @@ public Table table() {
       Dataset<Row> scanDF = cloneSession.read().format("iceberg")
           .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
           .option(SparkReadOptions.SPLIT_SIZE, splitSize(inputFileSize(filesToRewrite)))
+          .option(SparkReadOptions.LOOKBACK, 10)
```
is this needed? 10 is already the default

RussellSpitzer replied:
That's true ... probably fine to just remove it.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/BaseFileScanTask.java:
```diff
@@ -109,14 +110,11 @@ public String toString() {
 
-    OffsetsAwareTargetSplitSizeScanTaskIterator(
-        List<Long> offsetList, FileScanTask parentScanTask, long targetSplitSize) {
+    OffsetsAwareTargetSplitSizeScanTaskIterator(List<Long> offsetList, FileScanTask parentScanTask) {
```
Changing the behavior here to just always produce splits that are the size of a row group. Split-Size is not used here, only when recombining.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseFileScanTask.java:
```diff
@@ -224,5 +217,51 @@ public Expression residual() {
+
+    for (FileScanTask fileScanTask : tasks) {
+      if (!(fileScanTask instanceof SplitScanTask)) {
+        // We do not know how to combine anything but SplitScanTasks
```
Nit: personal pronouns don't usually add much to comments or documentation. It's unclear who "we" is and you can usually be more direct. Here I think it is sufficient to say "only SplitScanTasks produced by split are supported".

rdblue commented on core/src/main/java/org/apache/iceberg/BaseFileScanTask.java:
```diff
@@ -224,5 +217,51 @@ public Expression residual() {
+      combinedScans.add(lastSplit);
+    }
+
+    return combinedScans.stream().toArray(FileScanTask[]::new);
```
Since this method accepts `List<FileScanTask>`, I think it makes more sense to return a list as well. The caller can always convert to an array if that's appropriate.

RussellSpitzer replied:
Sounds good to me

rdblue commented on core/src/main/java/org/apache/iceberg/BaseFileScanTask.java:
```diff
@@ -224,5 +217,51 @@ public Expression residual() {
+    }
+  }
+
+  static FileScanTask[] combineSimilarTasks(List<FileScanTask> tasks) {
```
Using `Adjacent` rather than `Similar` is more specific, since this is combining adjacent splits with `isAdjacent`.

rdblue reviewed (approved):
This looks great to me. I like that it is simpler than the original approach and uses recombining. I had a few minor comments on the changes in `BaseFileScanTask`, but overall I think this is correct and ready to go.

I also took a quick look at updating bin packing and splitting so that bin packing can request a certain size split to avoid needing to recombine. That seems doable, but would require quite a few changes so I don't think it is worth it at this point.

RussellSpitzer commented:
All cleaned up based on your comments, @rdblue I'll merge when tests pass again unless you had other blockers.

RussellSpitzer commented:
getting some "works on my machine" errors, jenkins fails :( Will investigate
